File: prep/implementation/patterns_reference_imeplementation/sliding_window/FruitsIntoBaskets.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(1)
public class FruitsIntoBaskets {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        in.nextLine();

        String[] parts = in.nextLine().trim().split(" ");
        char[] a = new char[n];

        for (int i = 0; i < n; i++) {
          a[i] = parts[i].charAt(0);
        }

        Map<Character, Integer> freq = new HashMap<>();
        int maxLen = 0, windowStart = 0;

        for (int windowEnd = 0; windowEnd < n; windowEnd++) {
          char c = a[windowEnd];
          freq.put(c, freq.getOrDefault(c, 0) + 1);

          while (freq.size() > 2) {
            char d = a[windowStart];
            freq.put(d, freq.get(d) - 1);

            if (freq.get(d) == 0) {
              freq.remove(d);
            }

            windowStart++;
          }

          maxLen = Math.max(maxLen, windowEnd - windowStart + 1);
        }

        System.out.println(maxLen);
      }
    }
  }
}

File: prep/implementation/patterns_reference_imeplementation/sliding_window/fruits_into_baskets.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<char> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    int max_len = 0, window_start = 0;
    unordered_map<char, int> freq;

    for (int window_end = 0; window_end < n; window_end++) {
      char c = a[window_end];

      if (freq.find(c) == freq.end()) {
        freq[c] = 1;
      } else {
        freq[c]--;
      }

      while (freq.size() > 2) {
        char d = a[window_start];
        freq[d]--;

        if (freq[d] == 0) {
          freq.erase(d);
        }
        window_start++;
      }
      max_len = max(max_len, window_end - window_start + 1);
    }

    cout << max_len << "\n";
  }

  return 0;
}